File: client/src/lib/redact.test.ts
import { describe, expect, it } from "vitest";

import { REDACTED, redactSecrets, redactURL } from "./redact";

// A realistic session token: 32 random bytes, base64url (the shape
// server/internal/util/token mints).
const TOKEN = "Zk3q9_Rb-2xVn8LmPq4tYw7cHs1dJe6uKo0aBf5gTiA";
const INVITE = "iNv1te-T0ken_abcdefghijklmnop";

describe("redactSecrets — query parameters", () => {
  it("redacts ?token= in a ws URL and keeps game and player", () => {
    const url = `wss://cmd.example/ws?game=g-1&token=${TOKEN}&player=p-9`;
    expect(redactSecrets(url)).toBe(`wss://cmd.example/ws?game=g-1&token=${REDACTED}&player=p-9`);
  });

  it("redacts token as the first and the last parameter", () => {
    expect(redactSecrets(`/games/1/replay?token=${TOKEN}`)).toBe(
      `/games/1/replay?token=${REDACTED}`,
    );
    expect(redactSecrets(`/ws?token=${TOKEN}#frag`)).toBe(`/ws?token=${REDACTED}#frag`);
  });

  it("redacts the invite and reclaim ?t= in a hash route", () => {
    expect(redactSecrets(`https://h/#/games/abc/join?t=${INVITE}&spectator=1`)).toBe(
      `https://h/#/games/abc/join?t=${REDACTED}&spectator=1`,
    );
    expect(redactSecrets(`#/games/abc/reclaim?t=${INVITE}`)).toBe(
      `#/games/abc/reclaim?t=${REDACTED}`,
    );
  });

  it("redacts access_token, refresh_token, id_token and client_secret", () => {
    const got = redactSecrets(
      `cb?access_token=${TOKEN}&refresh_token=r1r1r1&id_token=i2&client_secret=s3&scope=identify`,
    );
    expect(got).toBe(
      `cb?access_token=${REDACTED}&refresh_token=${REDACTED}&id_token=${REDACTED}&client_secret=${REDACTED}&scope=identify`,
    );
  });

  it("redacts the oauth-complete fragment token", () => {
    expect(redactSecrets(`#/oauth-complete?token=${TOKEN}&expires_at=2026-09-17`)).toBe(
      `#/oauth-complete?token=${REDACTED}&expires_at=2026-09-17`,
    );
  });

  it("redacts a bare token= in prose", () => {
    expect(redactSecrets(`console.error: failed with token=${TOKEN} after retry`)).toBe(
      `console.error: failed with token=${REDACTED} after retry`,
    );
  });

  it("is case-insensitive in the key", () => {
    expect(redactSecrets(`?Token=${TOKEN}&ACCESS_TOKEN=x`)).toBe(
      `?Token=${REDACTED}&ACCESS_TOKEN=${REDACTED}`,
    );
  });

  it("redacts an HTML-escaped &amp;t=", () => {
    expect(redactSecrets(`join?game=1&amp;t=${INVITE}`)).toBe(`join?game=1&amp;t=${REDACTED}`);
  });
});

describe("redactSecrets — URL-encoded variants", () => {
  it("redacts a percent-encoded URL nested in another URL and keeps what follows", () => {
    const nested = encodeURIComponent(`/ws?token=${TOKEN}&game=g-1`);
    expect(nested).toContain("%3Ftoken%3D");
    const got = redactSecrets(`https://h/login?next=${nested}`);
    expect(got).not.toContain(TOKEN);
    expect(got).toBe(`https://h/login?next=%2Fws%3Ftoken%3D${REDACTED}%26game%3Dg-1`);
  });

  it("redacts an encoded invite ?t=", () => {
    const nested = encodeURIComponent(`#/games/abc/join?t=${INVITE}`);
    const got = redactSecrets(`redirect=${nested}`);
    expect(got).not.toContain(INVITE);
    expect(got).toContain(`t%3D${REDACTED}`);
  });

  it("redacts a token value that itself carries percent-escapes", () => {
    expect(redactSecrets("?token=ab%2Bcd%2Fef&game=1")).toBe(`?token=${REDACTED}&game=1`);
  });
});

describe("redactSecrets — headers and JSON", () => {
  it("redacts Authorization: Bearer", () => {
    expect(redactSecrets(`Authorization: Bearer ${TOKEN}`)).toBe(
      `Authorization: Bearer ${REDACTED}`,
    );
    expect(redactSecrets(`{"Authorization":"Bearer ${TOKEN}"}`)).toBe(
      `{"Authorization":"Bearer ${REDACTED}"}`,
    );
    expect(redactSecrets(`Authorization%3A%20Bearer%20${TOKEN}`)).toBe(
      `Authorization%3A%20Bearer%20${REDACTED}`,
    );
  });

  it("redacts JSON token fields, plain and escaped", () => {
    expect(redactSecrets(`{"token":"${TOKEN}","gameID":"g"}`)).toBe(
      `{"token":"${REDACTED}","gameID":"g"}`,
    );
    expect(redactSecrets(`"{\\"token\\": \\"${TOKEN}\\"}"`)).toBe(
      `"{\\"token\\": \\"${REDACTED}\\"}"`,
    );
  });

  it("leaves JSON booleans and card text alone", () => {
    const s = `{"token":true,"name":"Bearer of the Heavens","type_line":"Token Creature"}`;
    expect(redactSecrets(s)).toBe(s);
  });
});

describe("redactSecrets — false positives", () => {
  it("leaves ordinary log lines untouched", () => {
    for (const line of [
      "server error code=bad_request message=not your priority",
      "action create_token id=1234abcd",
      "snapshot seq=12 turn=3 step=draw",
      'chat from=Ada text="at t=3 I attacked"',
      "reconnecting in 500ms (attempt 2)",
      "card art failed to load: id=abc face=0 size=small",
      "wss://cmd.example/ws?game=g-1&player=p-9",
    ]) {
      expect(redactSecrets(line)).toBe(line);
    }
  });

  it("is idempotent", () => {
    const once = redactSecrets(`/ws?token=${TOKEN}&t=${INVITE}`);
    expect(redactSecrets(once)).toBe(once);
  });

  it("copes with empty input", () => {
    expect(redactSecrets("")).toBe("");
  });
});

describe("redactURL", () => {
  it("keeps host, path, game and player; redacts the token", () => {
    expect(redactURL(`ws://localhost:8080/ws?game=g&token=${TOKEN}&player=p`)).toBe(
      `ws://localhost:8080/ws?game=g&token=${REDACTED}&player=p`,
    );
  });

  it("does not throw on a malformed URL", () => {
    expect(redactURL(`::not a url?token=${TOKEN}`)).toBe(`::not a url?token=${REDACTED}`);
  });
});

File: server/internal/util/redact/redact_test.go
package redact

import (
	"net/url"
	"strings"
	"testing"
)

// token is a realistic session token: 32 random bytes, base64url, the
// shape internal/util/token mints.
const token = "Zk3q9_Rb-2xVn8LmPq4tYw7cHs1dJe6uKo0aBf5gTiA"

const invite = "iNv1te-T0ken_abcdefghijklmnop"

func TestSecretsRedacts(t *testing.T) {
	cases := []struct {
		name, in, want string
	}{
		{
			"ws URL keeps game and player",
			"connected to wss://cmd.example/ws?game=g-1&token=" + token + "&player=p-9",
			"connected to wss://cmd.example/ws?game=g-1&token=REDACTED&player=p-9",
		},
		{"token last", "/games/1/replay?token=" + token, "/games/1/replay?token=REDACTED"},
		{
			// auth.HMACAuthenticator's shape (#517): the dots must not
			// end the value, or the signed payload and signature leak.
			"signed session token",
			"/ws?token=v1.eyJyIjoicGxheWVyIn0.Zk3q9_Rb-2xVn8LmPq4tYw7cHs1dJe6uKo0aBf5gTiA&game=g-1",
			"/ws?token=REDACTED&game=g-1",
		},
		{
			"signed session bearer",
			"Authorization: Bearer v1.eyJyIjoicGxheWVyIn0.Zk3q9_Rb-2xVn8LmPq4tYw7cHs1dJe6uKo0aBf5gTiA",
			"Authorization: Bearer REDACTED",
		},
		{"token before fragment", "/ws?token=" + token + "#frag", "/ws?token=REDACTED#frag"},
		{
			"invite t in a hash route",
			"https://h/#/games/abc/join?t=" + invite + "&spectator=1",
			"https://h/#/games/abc/join?t=REDACTED&spectator=1",
		},
		{"reclaim t", "#/games/abc/reclaim?t=" + invite, "#/games/abc/reclaim?t=REDACTED"},
		{"html-escaped &amp;t", "join?game=1&amp;t=" + invite, "join?game=1&amp;t=REDACTED"},
		{
			"oauth tokens",
			"cb?access_token=" + token + "&refresh_token=r1&id_token=i2&client_secret=s3&scope=identify",
			"cb?access_token=REDACTED&refresh_token=REDACTED&id_token=REDACTED&client_secret=REDACTED&scope=identify",
		},
		{"bare token in prose", "failed with token=" + token + " after retry", "failed with token=REDACTED after retry"},
		{"case-insensitive key", "?Token=" + token + "&ACCESS_TOKEN=x", "?Token=REDACTED&ACCESS_TOKEN=REDACTED"},
		{"value with escapes", "?token=ab%2Bcd%2Fef&game=1", "?token=REDACTED&game=1"},
		{
			"url-encoded nested URL keeps trailing params",
			"https://h/login?next=" + url.QueryEscape("/ws?token="+token+"&game=g-1"),
			"https://h/login?next=%2Fws%3Ftoken%3DREDACTED%26game%3Dg-1",
		},
		{"double-encoded equals", "token%253D" + token, "token%253DREDACTED"},
		{"double-encoded equals, upper-case key and hex", "TOKEN%253d" + token, "TOKEN%253dREDACTED"},
		{
			"double-encoded nested URL keeps trailing params",
			"https://h/r?u=" + url.QueryEscape(url.QueryEscape("/ws?token="+token+"&game=g-1")),
			"https://h/r?u=%252Fws%253Ftoken%253DREDACTED%2526game%253Dg-1",
		},
		{
			"double-encoded invite t",
			"next=" + url.QueryEscape(url.QueryEscape("#/games/abc/join?t="+invite+"&x=1")),
			"next=%2523%252Fgames%252Fabc%252Fjoin%253Ft%253DREDACTED%2526x%253D1",
		},
		{"authorization bearer", "Authorization: Bearer " + token, "Authorization: Bearer REDACTED"},
		{"encoded bearer", "Authorization%3A%20Bearer%20" + token, "Authorization%3A%20Bearer%20REDACTED"},
		{"json token", `{"token":"` + token + `","gameID":"g"}`, `{"token":"REDACTED","gameID":"g"}`},
		{"escaped json token", `"{\"token\": \"` + token + `\"}"`, `"{\"token\": \"REDACTED\"}"`},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := Secrets(tc.in)
			if got != tc.want {
				t.Fatalf("Secrets(%q)\n got: %q\nwant: %q", tc.in, got, tc.want)
			}
			if strings.Contains(got, token) || strings.Contains(got, invite) {
				t.Fatalf("credential survived: %q", got)
			}
			if again := Secrets(got); again != got {
				t.Fatalf("not idempotent: %q -> %q", got, again)
			}
		})
	}
}

func TestSecretsEncodedInvite(t *testing.T) {
	got := Secrets("redirect=" + url.QueryEscape("#/games/abc/join?t="+invite))
	if strings.Contains(got, invite) || !strings.Contains(got, "t%3DREDACTED") {
		t.Fatalf("encoded invite not redacted: %q", got)
	}
}

// TestSecretsLeavesOrdinaryTextAlone pins the false positives that would
// make the redactor cost triage information.
func TestSecretsLeavesOrdinaryTextAlone(t *testing.T) {
	for _, s := range []string{
		"",
		"server error code=bad_request message=not your priority",
		"action create_token id=1234abcd",
		"snapshot seq=12 turn=3 step=draw",
		`chat from=Ada text="at t=3 I attacked"`,
		"card art failed to load: id=abc face=0 size=small",
		"wss://cmd.example/ws?game=g-1&player=p-9",
		`{"token":true,"name":"Bearer of the Heavens","type_line":"Token Creature"}`,
		"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36",
	} {
		if got := Secrets(s); got != s {
			t.Errorf("Secrets(%q) = %q, want unchanged", s, got)
		}
	}
}
